backends/generators/generator.py

```diff
@@ -0,0 +1,422 @@
+                continue
+
+            if data.get("kind") != "csr":
+                continue
+
+            found_csrs += 1
+            name = data.get("name")
+            if not name:
+                logging.error(f"Missing 'name' field in {path}")
+                continue
+
+            address = data.get("address")
+            indirect_address = data.get("indirect_address")
+
+            if not address and not indirect_address:
+                logging.error(
+                    f"Missing 'address' or 'indirect_address' field in CSR {name} in {path}"
+                )
+                address_errors += 1
+                continue
+
+            # Check if the CSR has a base constraint (32 or 64)
+            base = data.get("base")
+            if base:
+                if base == 32 and target_arch not in ["RV32", "BOTH"]:
+                    logging.debug(f"Skipping CSR {name} because it requires RV32 base")
+                    arch_filtered += 1
+                    continue
+                elif base == 64 and target_arch not in ["RV64", "BOTH"]:
+                    logging.debug(f"Skipping CSR {name} because it requires RV64 base")
+                    arch_filtered += 1
+                    continue
+
+            # If include_all is True, skip extension filtering
+            if not include_all:
+                # Check if this CSR is defined by an enabled extension
+                definedBy = data.get("definedBy")
+
+                # If definedBy is missing, log a warning but don't skip
+                # This is different from instructions where we're more strict
+                if definedBy is None:
+                    logging.warning(
+                        f"Missing 'definedBy' field in CSR {name} in {path}, including anyway"
+                    )
+                else:
+                    logging.debug(f"CSR {name} definedBy: {definedBy}")
+                    meets_extension_req = parse_extension_requirements(definedBy)
+                    if not meets_extension_req(enabled_extensions):
+                        msg = (
+                            f"Skipping CSR {name} because its extension is not enabled"
+                        )
+                        logging.debug(msg)
+                        extension_filtered += 1
+                        continue
+
+            # If we're here, we've passed all checks
+            try:
+                # Use address if available, otherwise use indirect_address
+                addr_to_use = address if address is not None else indirect_address
+                if isinstance(addr_to_use, int):
+                    addr_int = addr_to_use
+                else:
+                    addr_int = int(addr_to_use, 0)
+
+                # For BOTH architecture, add suffix to RV32-specific CSRs
+                if target_arch == "BOTH" and base == 32:
+                    csrs[addr_int] = f"{name.upper()}.RV32"
```

Comment:
This will eventually end up emitting invalid C code:
```
#define CSR_CYCLEH.RV32 0xc80
```

Preprocessor tokens cannot have ".". Maybe use underscore here?